eng/pipelines/templates/steps/sync-repo-branch.yml

```diff
@@ -1,17 +1,102 @@
+  #     Branch: master
+  #     TargetRepos: 
+  #       Azure/azure-sdk-for-go-pr:
+  - name: Repos
+    type: object
+  # Github PAT
+  - name: GH_TOKEN
+    type: string
+  # Github alias for git config
+  - name: GithubAlias
+    type: string
+    default: azure-sdk
+
 
 steps:
 
+- pwsh: |
+    git config --global user.email "${{ parameters.GithubAlias }}@microsoft.com"
+    git config --global user.name "${{ parameters.GithubAlias }}"
+  displayName: Init git config
+
 - ${{ each repo in parameters.Repos }}:
-  - script: |
-      git clone --filter=blob:none --branch ${{ repo.value.Branch }} https://github.com/${{ repo.key }} ${{ repo.key }}
+  - pwsh: |
+      Set-PsDebug -Trace 1
+      $SourceRepo = '${{ repo.key }}'
+      $SourceBranch = '${{ repo.value.Branch }}'
+
+      if (-not (Test-Path ${{ repo.key }})) {
+        New-Item -Path ${{ repo.key }} -ItemType Directory -Force
+        Set-Location $SourceRepo
+        git init
+        git remote add Source "https://${{ parameters.GH_TOKEN }}@github.com/${SourceRepo}.git"
+      } else {
+        Set-Location $SourceRepo
+      }
+      git fetch --no-tags Source $SourceBranch
+      if ($LASTEXITCODE -eq 0) {
+        git checkout -B SourceBranch "refs/remotes/Source/${SourceBranch}"
+      } else {
+        Write-Error "#`#vso[task.logissue type=error]Failed to fetch ${SourceRepo}/${SourceBranch}"
+      }
+
+      Set-PsDebug -Off
+
     displayName: ${{ repo.key }} - Clone ${{ repo.value.Branch }}
     continueOnError: true
 
   - ${{ each target in repo.value.TargetRepos }}:
-    - script: | 
-        git push https://$(azuresdk-github-pat)@github.com/${{ target.key }} ${{ coalesce(target.value.Branch, repo.value.Branch) }}
+    - pwsh: |
+        Set-PsDebug -Trace 1
+        $SourceBranch = '${{ repo.value.Branch }}'
+        $TargetRepo = '${{ target.key }}'
+        $TargetBranch = '${{ coalesce(target.value.Branch, repo.value.Branch) }}'
+
+        try {
+          git remote add Target "https://${{ parameters.GH_TOKEN }}@github.com/${TargetRepo}.git"
+
+          if (-not '${{ target.value.Rebase }}') {
+            git checkout -B TargetBranch SourceBranch
+          } else {
+            git fetch --no-tags Target $TargetBranch
+            if ($LASTEXITCODE -eq 0) {
+              git checkout -B TargetBranch "refs/remotes/Target/${TargetBranch}"
+            } else {
+              Write-Error "#`#vso[task.logissue type=error]Failed to fetch TargetBranch ${TargetBranch}."
+              Return
+            }
+            
+            git rebase --strategy-option=theirs SourceBranch
+            if ($LASTEXITCODE -ne 0) {
+              Write-Error "#`#vso[task.logissue type=error]Failed to rebase for ${TargetRepo}/${TargetBranch}"
+              git status
+              git diff
+              git rebase --abort
+              Return
+            }
+          }
+
+          git push --force Target "TargetBranch:refs/heads/${TargetBranch}"
```

Comment:
I don't like use always doing a force push. I know we might need to do that for the rebase case but for the non-rebase case we shouldn't do a force push. We should actually split the logic for rebase and non-rebase.

Reply:
Hummm makes sense

Reply:
Resolved